rust/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+pub mod dalek;
+
+#[cfg(feature = "openssl")]
+pub mod openssl;
+
+/// The [tests] module provides functions to test implementations of [PrivateKey]/[PublicKey].
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+    use crate::PrivateKey;
+    use hex_literal::hex;
+
+    pub fn run_vector_1<T: PrivateKey>(load_keypair: &dyn Fn([u8; 32]) -> (T, T::PublicKey)) {
+        let (apriv, apub) = load_keypair(hex!(
+            "77076d0a7318a57d3c16c17251b26645df4c2f87ebc0992ab177fba51db92c2a"
+        ));
+        let (bpriv, bpub) = load_keypair(hex!(
+            "5dab087e624a8a4b79e17f8b83800ee66f3bb1292618b6fd1c2f8b27ff88e0eb"
+        ));
+        let expected = hex!("9c44389f462d35d0672faf73a5e118f8b9f5c340bbe8d340e2b947c205ea4fa3");
+        assert_eq!(tag(&apriv, &bpub, 0, b"The quick brown fox"), expected);
+        assert_ne!(tag(&bpriv, &apub, 0, b"The quick brown fox"), expected);
+        assert_eq!(
+            verify(
+                &bpriv,
+                &apub,
+                0,
+                b"The quick brown fox",
+                &hex!("9c44389f462d")
+            ),
+            true
+        );
+        assert_eq!(
+            verify(
+                &bpriv,
+                &apub,
+                0,
+                b"The quick brown fox",
+                &hex!("ffeeddccbbaa")
+            ),
+            false
+        );
+        assert_eq!(
+            verify(
+                &apriv,
+                &bpub,
+                0,
+                b"The quick brown fox",
+                &hex!("9c44389f462d")
+            ),
+            false
+        );
+    }
+
+    pub fn run_vector_2<T: PrivateKey>(load_keypair: &dyn Fn([u8; 32]) -> (T, T::PublicKey)) {
+        let (apriv, apub) = load_keypair(hex!(
+            "fee1deadfee1deadfee1deadfee1deadfee1deadfee1deadfee1deadfee1dead"
+        ));
+        let (bpriv, bpub) = load_keypair(hex!(
+            "b105f00db105f00db105f00db105f00db105f00db105f00db105f00db105f00d"
+        ));
+        let expected = hex!("06476f1f314b06c7f96e5dc62b2308268cbdb6140aefeeb55940731863032277");
+        assert_eq!(tag(&bpriv, &apub, 100, b"The quick brown fox"), expected);
+        assert_ne!(tag(&apriv, &bpub, 100, b"The quick brown fox"), expected);
+        assert_eq!(
+            verify(
+                &apriv,
+                &bpub,
+                100,
+                b"The quick brown fox",
+                &hex!("06476f1f314b")
+            ),
+            true
+        );
+        assert_eq!(
+            verify(
+                &apriv,
+                &bpub,
+                100,
+                b"The quick brown fox",
+                &hex!("ffeeddccbbaa")
+            ),
+            false
+        );
+        assert_eq!(
+            verify(
+                &bpriv,
+                &apub,
+                100,
+                b"The quick brown fox",
+                &hex!("06476f1f314b")
+            ),
+            false
+        );
```

Comment:
`cargo clippy` warnings:
```suggestion
        assert!(verify(
            &apriv,
            &bpub,
            100,
            b"The quick brown fox",
            &hex!("06476f1f314b")
        ));
        assert!(!verify(
            &apriv,
            &bpub,
            100,
            b"The quick brown fox",
            &hex!("ffeeddccbbaa")
        ));
        assert!(!verify(
            &bpriv,
            &apub,
            100,
            b"The quick brown fox",
            &hex!("06476f1f314b")
        ));
```